### PR TITLE
First Delivery/MiguelGrover

### PR DESCRIPTION
**Pull request by MiguelAGrover**
The first delivery is contained inside main.go
Architecture that is going to be used: Clean architecture
Todo of tasks remaining for the final delivery can be found on README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -128,3 +130,18 @@ For submitting your work, you should follow these steps:
 - [Packages](https://www.youtube.com/watch?v=sf7f4QGkwfE)
 - [Unit testing](https://golang.org/pkg/testing/)
 - [Functions as values](https://tour.golang.org/moretypes/24)
+
+</details>
+
+### Todos
```
Comment (alvarocueva) [inline README.md:136]: Add an empty line

File: README.md [hunk 2]
```diff
@@ -128,3 +130,18 @@ For submitting your work, you should follow these steps:
...
+- [x] First Delivery
+- [ ] Change config setup to allow csv file names and remove database connections
+- [ ] Consider gormcsv to manipulate the csv.
+- [ ] Add more routes to router and it's logic.
+- [ ] Add remaining comments to procedures from the architecture
+- [ ]  Add unit testcases
```
Comment (alvarocueva) [inline README.md:142]: Remove the extra space before `Add`

File: README.md [hunk 3]
```diff
@@ -128,3 +130,18 @@ For submitting your work, you should follow these steps:
...
+
+### Method of running
+
+- go build .
+- run the generated digimons file.
```
Comment (alvarocueva) [inline README.md:147]: Add an empty line at the end of the file

File: config/config.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package config
+
+import (
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+
+	"github.com/davecgh/go-spew/spew"
+
```
Comment (alvarocueva) [inline config/config.go:10]: Remove this line

File: config/config.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+  password: root
+  net: tcp
+  addr: localhost:3306
+  dbname: golang_bootcamp_2020
+  allowNativePasswords: true
+  params:
+    parseTime: "true"
+
+server:
+  address: 8080
```
Comment (alvarocueva) [inline config/config.yml:12]: By convention, this kind of files should end with an empty line

File: domain/model/Digimon.go [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package model
+
+// Digimon is the strcuture to contain digimon data.
```
Comment (alvarocueva) [inline domain/model/Digimon.go:3]: `strcuture` -> `structure`

File: infraestructure/datastore/db.go [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package datastore
+
+import (
+	"log"
+
+	// "github.com/jinzhu/gorm"
```
Comment (alvarocueva) [inline infraestructure/datastore/db.go:6]: Remove this

File: infraestructure/datastore/db.go [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+)
+
+// NewDB returns the instance of the connection to sqlite
+func NewDB() *gorm.DB {
+	DBName := "config/goolang-bootcamp-2020.db"
```
Comment (alvarocueva) [inline infraestructure/datastore/db.go:13]: You could add this file path to the config file and, in the main, pass the value as a parameter to use it here.

File: interface/controller/app_controller.go [hunk 9]
```diff
@@ -0,0 +1,5 @@
+package controller
+
+type AppController struct {
```
Comment (alvarocueva) [inline interface/controller/app_controller.go:3]: Add a description for this exported function

File: interface/controller/context.go [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package controller
+
+type Context interface {
```
Comment (alvarocueva) [inline interface/controller/context.go:3]: Add a description for this exported function

File: interface/controller/digimon_controller.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+	"digimons/domain/model"
+	"digimons/usecase/interactor"
+)
+
+type digimonController struct {
+	digimonController interactor.DigimonInteractor
+}
+
+type DigimonController interface {
```
Comment (alvarocueva) [inline interface/controller/digimon_controller.go:14]: Add a description for this exported interface

File: interface/controller/digimon_controller.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	GetDigimons(c Context) error
+}
+
+func NewDigimonController(di interactor.DigimonInteractor) DigimonController {
```
Comment (alvarocueva) [inline interface/controller/digimon_controller.go:18]: Add a description for this exported function

File: interface/controller/digimon_controller.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+	return &digimonController{di}
+}
+
+func (di *digimonController) GetDigimons(c Context) error {
```
Comment (alvarocueva) [inline interface/controller/digimon_controller.go:22]: Add a description for this exported function

File: interface/controller/digimon_controller.go [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+	var d []*model.Digimon
+
+	d, err := di.digimonController.Get(d)
+	if err != nil {
+		return err
```
Comment (alvarocueva) [inline interface/controller/digimon_controller.go:27]: Why not 
```go
c.JSON(http.StatusNotFound, d)
```
Or something like that

File: interface/presenter/digimon_presenter.go [hunk 15]
```diff
@@ -0,0 +1,22 @@
+package presenter
+
+import "digimons/domain/model"
+
+type digimonPresenter struct {
```
Comment (alvarocueva) [inline interface/presenter/digimon_presenter.go:5]: One line
```go
type digimonPresenter struct{}
```

File: interface/presenter/digimon_presenter.go [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+type DigimonPresenter interface {
```
Comment (alvarocueva) [inline interface/presenter/digimon_presenter.go:8]: Add a description for this exported interface

File: interface/presenter/digimon_presenter.go [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+	ResponseDigimons(di []*model.Digimon) []*model.Digimon
+}
+
+func NewDigimonPresenter() DigimonPresenter {
```
Comment (alvarocueva) [inline interface/presenter/digimon_presenter.go:12]: Add a description for this exported function

File: interface/presenter/digimon_presenter.go [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+	return &digimonPresenter{}
+}
+
+func (dp *digimonPresenter) ResponseDigimons(ds []*model.Digimon) []*model.Digimon {
```
Comment (alvarocueva) [inline interface/presenter/digimon_presenter.go:16]: Add a description for this exported function

File: interface/presenter/digimon_presenter.go [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+	for _, d := range ds {
+		d.Name = d.Name
```
Comment (alvarocueva) [inline interface/presenter/digimon_presenter.go:18]: Why?
This is a self-assignment, is it what you wanted to do?

File: interface/repository/digimon_repository.go [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+	"digimons/domain/model"
+
+	"gorm.io/gorm"
+)
+
+type digimonRepository struct {
+	db *gorm.DB
+}
+
+type DigimonRepository interface {
```
Comment (alvarocueva) [inline interface/repository/digimon_repository.go:13]: Add a description for this exported interface

File: interface/repository/digimon_repository.go [hunk 21]
```diff
@@ -0,0 +1,29 @@
...
+	FindAll(d []*model.Digimon) ([]*model.Digimon, error)
+}
+
+func NewDigimonRepository(db *gorm.DB) DigimonRepository {
```
Comment (alvarocueva) [inline interface/repository/digimon_repository.go:17]: Add a description for this exported function

File: interface/repository/digimon_repository.go [hunk 22]
```diff
@@ -0,0 +1,29 @@
...
+	return &digimonRepository{db}
+}
+
+func (dr *digimonRepository) FindAll(d []*model.Digimon) ([]*model.Digimon, error) {
```
Comment (alvarocueva) [inline interface/repository/digimon_repository.go:21]: Add a description for this exported function

File: registry/digimon_registry.go [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+import (
+	"digimons/interface/controller"
+	ip "digimons/interface/presenter"
+	ir "digimons/interface/repository"
+	"digimons/usecase/interactor"
+	up "digimons/usecase/presenter"
+	ur "digimons/usecase/repository"
+)
+
+func (r *registry) NewDigimonController() controller.DigimonController {
```
Comment (alvarocueva) [inline registry/digimon_registry.go:12]: All the functions here should have a description since they are exported

File: main.go [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+	}
+
+	defer sqlDB.Close()
+
+	r := registry.NewRegistry(db)
+
+	e := echo.New()
+	e = router.NewRouter(e, r.NewAppController())
+
+	fmt.Println("Server listen at http://localhost" + ":" + config.C.Server.Address)
```
Comment (alvarocueva) [inline main.go:43]: This print should be moved below the err validation, right?

File: main.go [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+	if err := e.Start(":" + config.C.Server.Address); err != nil {
+		log.Fatalln(err)
+	}
+}
+
+// write Obtain data from an external API convert it to an array and save it into csv file
+func write() {
+	resp, err := http.Get("https://digimon-api.vercel.app/api/digimon")
```
Comment (alvarocueva) [inline main.go:51]: You could set this host in the config file

File: main.go [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	defer resp.Body.Close()
+	bodyBytes, _ := ioutil.ReadAll(resp.Body)
```
Comment (alvarocueva) [inline main.go:57]: You should handle the possible error here

File: main.go [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+	fmt.Println(bodyBytes)
+
+	var DigimonStructArray []model.Digimon
+	json.Unmarshal(bodyBytes, &DigimonStructArray)
+
+	csvFile, err := os.Create("./config/digimon.csv")
```
Comment (alvarocueva) [inline main.go:63]: You could get this path from the config file

File: main.go [hunk 28]
```diff
@@ -0,0 +1,111 @@
...
+	for _, digimon := range DigimonStructArray {
+		var row []string
+		row = append(row, digimon.Name)
+		row = append(row, digimon.Level)
+		row = append(row, digimon.Image)
+		writer.Write(row)
+	}
+	// remember to flush!
+	writer.Flush()
+
```
Comment (alvarocueva) [inline main.go:79]: Remove this line

File: registry/registry.go [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+	"digimons/interface/controller"
+
+	"gorm.io/gorm"
+)
+
+type registry struct {
+	db *gorm.DB
+}
+
+type Registry interface {
```
Comment (alvarocueva) [inline registry/registry.go:13]: The exported interface and functions should have description

File: usecase/interactor/digimon_interactor.go [hunk 30]
```diff
@@ -0,0 +1,29 @@
...
+	"digimons/usecase/presenter"
+	"digimons/usecase/repository"
+)
+
+type digimonInteractor struct {
+	DigimonRepository repository.DigimonRepository
+	DigimonPresenter  presenter.DigimonPresenter
+}
+
+type DigimonInteractor interface {
```
Comment (alvarocueva) [inline usecase/interactor/digimon_interactor.go:14]: The exported interface and functions should have description

File: usecase/presenter/digimon_presenter.go [hunk 31]
```diff
@@ -0,0 +1,7 @@
+package presenter
+
+import "digimons/domain/model"
+
+type DigimonPresenter interface {
```
Comment (alvarocueva) [inline usecase/presenter/digimon_presenter.go:5]: Add a description for this exported interface

File: usecase/repository/digimon_repository.go [hunk 32]
```diff
@@ -0,0 +1,7 @@
+package repository
+
+import "digimons/domain/model"
+
+type DigimonRepository interface {
```
Comment (alvarocueva) [inline usecase/repository/digimon_repository.go:5]: Add a description for this exported interface

**Comment by MiguelAGrover** [comment]
I will make the PR on my fork as it was expected. Thanks.

**Pull request closed without merging**